src/element/builders.rs

```diff
@@ -1,83 +1,64 @@
 ///     .push(1)
 ///     .push(true)
 ///     .push("foo")
-///     .build()
-///     .into();
-/// let expected = Element::read_one(r#"[1, true, "foo"]"#).unwrap();
+///     .build();
+/// let expected: Sequence = Sequence::new([
+///   Element::integer(1),
+///   Element::boolean(true),
+///   Element::string("foo"),
+/// ]);
 /// assert_eq!(actual, expected);
 /// ```
-pub struct ListBuilder {
-    values: Vec<Element>,
-}
-
-impl ListBuilder {
-    /// Crate visible; users should call [`List::builder()`] or [`Element::list_builder()`] instead.
-    pub(crate) fn new() -> Self {
-        Self { values: Vec::new() }
-    }
-
-    /// Helper method for [`List::clone_builder()`].
-    pub(crate) fn with_initial_elements(elements: &[Element]) -> Self {
-        let mut new_elements = Vec::with_capacity(elements.len());
-        new_elements.extend_from_slice(elements);
-        Self {
-            values: new_elements,
-        }
-    }
-
-    /// Adds the provided element to the end of the [`List`] being constructed.
-    pub fn push<E: Into<Element>>(mut self, element: E) -> Self {
-        self.values.push(element.into());
-        self
-    }
-
-    /// Removes the element at the specified position from the [`List`] being constructed.
-    /// If the index is out of bounds, this method will panic.
-    pub fn remove(mut self, index: usize) -> Self {
-        // This has O(n) behavior; the removals could be
-        // buffered until the build() if needed.
-        self.values.remove(index);
-        self
-    }
-
-    /// Builds a [`List`] with the previously specified elements.
-    pub fn build(self) -> List {
-        List::new(self.values)
-    }
-}
-
-/// Constructs [SExp] values incrementally.
-///
+/// Building a [List]:
 /// ```
-/// use ion_rs::element::Element;
-/// let actual: Element = Element::sexp_builder()
+/// use ion_rs::element::{Element, List, Sequence};
+/// let actual: List = Sequence::builder()
 ///     .push(1)
 ///     .push(true)
 ///     .push("foo")
-///     .build()
-///     .into();
-/// let expected = Element::read_one(r#"(1 true "foo")"#).unwrap();
+///     .build_list();
+/// let expected: List = List(Sequence::new([
+///   Element::integer(1),
+///   Element::boolean(true),
+///   Element::string("foo"),
+/// ]));
+/// assert_eq!(actual, expected);
+/// ```
+/// Building a [SExp]:
+/// ```
+/// use ion_rs::element::{Element, SExp, Sequence};
+/// let actual: SExp = Sequence::builder()
+///     .push(1)
+///     .push(true)
+///     .push("foo")
+///     .build_sexp();
+/// let expected: SExp = SExp(Sequence::new([
+///   Element::integer(1),
+///   Element::boolean(true),
+///   Element::string("foo"),
+/// ]));
 /// assert_eq!(actual, expected);
 /// ```
-pub struct SExpBuilder {
+pub struct SequenceBuilder {
     values: Vec<Element>,
 }
 
-impl SExpBuilder {
-    /// Crate visible; users should call [`SExp::builder()`] or [`Element::sexp_builder()`] instead.
+impl SequenceBuilder {
```

Comment:
If we parameterize `SequenceBuilder`, the return type of `build()` can be set when the SequenceBuilder is created so that you can't write `List::builder().build_sexp()`.
```rust
struct SequenceBuilder<T> {
    _target_type: Phantom<T>,
    values: Vec<Element>,
}

impl <T> SequenceBuilder<T> {
    /// Crate visible; users should call [`Sequence::builder()`] instead.
    pub(crate) fn new() -> Self {
        Self { values: Vec::new() }
    }

    /// Helper method for [`Sequence::clone_builder()`].
    pub(crate) fn with_initial_elements(elements: &[Element]) -> Self {
        let mut new_elements = Vec::with_capacity(elements.len());
        new_elements.extend_from_slice(elements);
        Self {
            values: new_elements,
        }
    }

    // etc.
}

// Repeate for Sequence and SExp
impl SequenceBuilder<List> {
    pub fn build(self) -> List {
        List(self.values.into())
    }
}
```

Reply:
This would work, though I think my current leaning would be to simply eliminate the `List::builder() -> SequenceBuilder` and `SExp::builder() -> SequenceBuilder` methods in favor of the (already available) `Sequence::builder()`.

The doc comment examples for `List` and `SExp` can demonstrate using `Sequence::builder()` to construct them, and most users will reach for the macros anyway.

An interesting (minor) benefit of using the same builder type for all sequences is that I can do:
```rust
let sexp = ion_sexp!(1 2 3);
// I can make a List using the SExp's elements as a starting point
let list = sexp.clone_builder().build_list();
```

It's possible that we'll want to introduce a `Stream` sequence type--values with no container, which are used at the top level and (in Ion 1.1) template expansion. We would also be able to use `SequenceBuilder` for that.

Reply:
> An interesting (minor) benefit of using the same builder type for all sequences is that I can do:
> ```rust
> let list = sexp.clone_builder().build_list();
> ```

We could use the generics so that `List::builder()` only lets you build a list, but `list.clone_builder()` will let you build a `List`, `SExp`, or `Sequence`. E.g.:
```rust
impl SequenceBuilder<Cloned> {
    pub fn build_list(self) -> List {
        List(self.values.into())
    }
    pub fn build_sexp(self) -> SExp {
        List(self.values.into())
    }
    // Etc.
}
```

That being said, I think the solution you've proposed is fine (and certainly simpler).